api/api_request_logger.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	log "github.com/inconshreveable/log15"
+)
+
+// RequestLogger logs requests to an output
+type RequestLogger struct {
+	enabled      bool
+	writerChan   chan entry
+	stopChan     chan bool
+	outputWriter func(msg string, ctx ...interface{})
+}
+
+// NewThorRequestLogger applies thor business logic
+// when returning an instance of *RequestLogger
+func NewThorRequestLogger(enabledFlag bool, enableLog bool) *RequestLogger {
+	if enabledFlag {
+		return NewRequestLogger(enabledFlag, log.Info)
+	}
+	if enableLog {
+		return NewRequestLogger(enableLog, log.Debug)
+	}
+	return NewRequestLogger(false, nil)
+}
+
+func NewRequestLogger(enabled bool, logger func(msg string, ctx ...interface{})) *RequestLogger {
+	return &RequestLogger{
+		enabled:      enabled,
+		outputWriter: logger,
+		writerChan:   make(chan entry, 100_000),
+		stopChan:     make(chan bool),
+	}
+}
+
+func (l *RequestLogger) Enabled() bool {
+	return l.enabled
+}
+
+// Handle returns an http handler to ensure requests are syphoned into the writer
+func (l *RequestLogger) Handle(handler http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		// Read and log the body (note: this can only be done once)
+		// Ensure you don't disrupt the request body for handlers that need to read it
+		var bodyBytes []byte
+		var err error
+		if r.Body != nil {
+			bodyBytes, err = io.ReadAll(r.Body)
+			if err != nil {
+				log.Warn("unexpected body read error", "err", err)
+				return // don't pass bad request to the next handler
+			}
+			r.Body = io.NopCloser(io.Reader(bytes.NewReader(bodyBytes)))
+		}
+
+		l.writerChan <- entry{
+			Timestamp: time.Now(),
+			URI:       r.URL.String(),
+			Method:    r.Method,
+			Body:      string(bodyBytes),
+		}
+
+		// call the original http.Handler we're wrapping
+		handler.ServeHTTP(w, r)
+	}
+
+	// start the writer
+	l.start()
+
+	// http.HandlerFunc wraps a function so that it
+	// implements http.Handler interface
+	return http.HandlerFunc(fn)
+}
+
+func (l *RequestLogger) Stop() {
+	if !l.enabled {
+		return
+	}
+
+	// make sure any pending message is written
+	l.stopChan <- true
+	<-l.stopChan
+}
+
+func (l *RequestLogger) start() {
+	go func() {
+		for {
+			select {
+			case item := <-l.writerChan:
+				marshal, err := json.Marshal(item)
+				if err != nil {
+					log.Warn("unable to marshal api request entry", "err", err)
+					continue
+				}
+
+				l.outputWriter("[request] " + string(marshal))
+			case <-l.stopChan:
+				close(l.stopChan)
+				return
+			}
+		}
+	}()
+}
+
+type entry struct {
+	Timestamp time.Time `json:"timestamp"`
```

Comment:
The timestamp is there in the log, seems redundant
```
INFO[05-17|15:23:46] [request] {"timestamp":"2024-05-17T15:23:46.036722+08:00","uri":"/accounts/*","method":"POST","body":"{\n  \"gas\": 50000,\n  \"gasPrice\": \"1000000000000000\",\n  \"caller\": \"0x6d95e6dca01d109882fe1726a2fb9865fa41e7aa\",\n  \"provedWork\": \"1000\",\n  \"gasPayer\": \"0xd3ae78222beadb038203be21ed5ce7c9b1bff602\",\n  \"expiration\": 1000,\n  \"blockRef\": \"0x00000000851caf3c\",\n  \"clauses\": [\n    {\n      \"to\": \"0x0000000000000000000000000000456E65726779\",\n      \"value\": \"0x0\",\n      \"data\": \"0xa9059cbb0000000000000000000000000f872421dc479f3c11edd89512731814d0598db50000000000000000000000000000000000000000000000013f306a2409fc0000\"\n    },\n    {\n      \"to\": \"0xf077b491b355E64048cE21E3A6Fc4751eEeA77fa\",\n      \"value\": \"0x6124fee993bc00000\",\n      \"data\": \"0x\"\n    },\n    {\n      \"to\": null,\n      \"value\": \"0x0\",\n      \"data\": \"0x6080604052348015600f57600080fd5b50609f8061001e6000396000f300608060405260043610603f576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680631820cabb146044575b600080fd5b348015604f57600080fd5b506056606c565b6040518082815260200191505060405180910390f35b62015180815600a165627a7a723058200ac7475da248e2fc26c057319e296e90c24d5f8b9bf956fb3b77545642cad3b10029\"\n    }\n  ]\n}"} ```

Reply:
Understood, however the timestamp of the log would need to be more accurate, and there might be differences between the event time and the time when it was logged.